src/backend/cuda/cast.hpp

```diff
@@ -89,9 +89,40 @@ struct CastWrapper {
     Array<To> operator()(const Array<Ti> &in) {
         CastOp<To, Ti> cop;
         common::Node_ptr in_node = in.getNode();
-        common::UnaryNode *node  = new common::UnaryNode(
-            static_cast<af::dtype>(dtype_traits<To>::af_type), cop.name(),
-            in_node, af_cast_t);
+        af::dtype to_dtype = static_cast<af::dtype>(dtype_traits<To>::af_type);
+
+        // JIT optimization in the cast of multiple sequential casts that become
+        // idempotent - check to see if the previous operation was also a cast
+        // TODO: handle arbitrarily long chains of casts
+        auto in_node_nary =
+            std::dynamic_pointer_cast<common::NaryNode>(in_node);
+        if (in_node_nary && in_node_nary->getOp() == af_cast_t) {
+            // The only way to get the input type of the child node if it's a
+            // cast is to get the output type of the child's child.
+            auto in_node_children = in_node_nary->getChildren();
+            // Check if any children are casts with the same type - if so,
+            // insert a shortcut noop
+            for (size_t i = 0; i < in_node_children.size(); ++i) {
+                // Found a node whose child can be fast-track noop-ed
+                common::Node_ptr in_in_node = in_node_nary->getChildren()[i];
+                if (in_in_node->getType() == to_dtype) {
+                    // If the output of the child's child is the same as the
+                    // output of this node, ignore the input node and simply
+                    // connect a noop node from the child's child to produce
+                    // this op's output
+
+                    // TODO: including unary.hpp to use
+                    // detail::unaryName<af_noop_t> breaks some other stuff
+                    return createNodeArray<To>(
```

Comment:
Instead of creating a noop node. Is it possible to just pass the in_in_node ptr to the createNodeArray? That way you don't need to call a new kernel to do the noop step.

Reply:
If you're referring to `return createNodeArray<To>(in.dims(), in_node);`, that doesn't seem to work — the output isn't equal to the input. I'll have to read into these functions to understand why that's the case - do you have any idea?